### PR TITLE
7 post /api/articles/article id/comments

### PR DESCRIPTION
**Pull request by jaketysome**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
         });
     });
   });
+  describe.only("POST/api/articles/:article_id/comments", () => {
```
Comment (popatre) [inline __tests__/app.test.js]: Remove .only before submitting, otherwise its not running all the tests

File: __tests__/app.test.js [hunk 2]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+        .expect(201)
+        .then((response) => {
+          const postedComment = response.body.postedComment;
+          expect(postedComment.article_id).toBe(2);
+          expect(postedComment).toEqual(
+            expect.objectContaining({
+              comment_id: expect.any(Number),
+              body: expect.any(String),
+              article_id: expect.any(Number),
+              author: expect.any(String),
```
Comment (popatre) [inline __tests__/app.test.js]: You need to check the author is the one your sent in the post body

File: __tests__/app.test.js [hunk 3]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+        .post("/api/articles/2/comments")
+        .send(requestBody)
+        .expect(201)
+        .then((response) => {
+          const postedComment = response.body.postedComment;
+          expect(postedComment.article_id).toBe(2);
+          expect(postedComment).toEqual(
+            expect.objectContaining({
+              comment_id: expect.any(Number),
+              body: expect.any(String),
```
Comment (popatre) [inline __tests__/app.test.js]: You need to check the body is the one you sent, not just any

File: __tests__/app.test.js [hunk 4]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+              article_id: expect.any(Number),
```
Comment (popatre) [inline __tests__/app.test.js]: You can check the article_id here, without doing a separate assertion

Comment (popatre) [reply __tests__/app.test.js]: Like the one on line 189

File: __tests__/app.test.js [hunk 5]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+        username: "cool_guy73",
+        body: "generic comment",
+      };
+      return request(app)
+        .post("/api/articles/4/comments")
+        .send(requestBody)
+        .expect(400)
+        .then((response) => {
+          const msg = response.body.msg;
+          expect(msg).toBe("bad request");
```
Comment (popatre) [inline __tests__/app.test.js]: This needs to be 404 - the request is valid, but the username is not found in the DB

File: __tests__/app.test.js [hunk 6]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+        });
+    });
+    test("status: 400, missing required request keys", () => {
+      const requestBody = {
+        body: "generic comment",
```
Comment (popatre) [inline __tests__/app.test.js]: I would change this to either be an empty object, or the username is there, but the body key is not

File: __tests__/app.test.js [hunk 7]
```diff
@@ -174,4 +174,80 @@ describe("api", () => {
...
+      };
+      return request(app)
+        .post("/api/articles/2/comments")
+        .send(requestBody)
+        .expect(400)
+        .then((response) => {
+          const msg = response.body.msg;
+          expect(msg).toBe("bad request");
+        });
+    });
```
Comment (popatre) [inline __tests__/app.test.js:250]: Another test you could add:
the keys are there on the post body, but the data type is wrong - for example is expects a string in the body, but you send a number

File: controllers/controllers.articles.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ const {
   selectArticleById,
   selectCommentsByArticleId,
   checkIfArticleExists,
+  insertCommentByArticleId,
```
Comment (popatre) [inline controllers/controllers.articles.js:7]: good naming

File: controllers/controllers.articles.js [hunk 9]
```diff
@@ -44,21 +45,36 @@ exports.getArticleById = (req, res, next) => {
...
-      next(err);
-    });
+    .catch(next);
+};
+
+exports.postCommentByArticleId = (req, res, next) => {
+  const articleId = req.params.article_id;
+  const reqBody = req.body;
+
+  const promises = [
```
Comment (popatre) [inline controllers/controllers.articles.js]: nice naming again

File: controllers/controllers.articles.js [hunk 10]
```diff
@@ -44,21 +45,36 @@ exports.getArticleById = (req, res, next) => {
...
+    checkIfArticleExists(articleId),
+    insertCommentByArticleId(articleId, reqBody),
+  ];
+
+  Promise.all(promises)
+    .then(([articleExists, postedComment]) => {
```
Comment (popatre) [inline controllers/controllers.articles.js]: In terms of good prac - we don't want unused variables if possible - it just fills memory space for no reason. 
I dont think articleExists is used anywhere. 
Does this still work if you switch the order of the promises above? Then you could just use the postedComment, and wouldnt need to even have the articleExists variable there

File: controllers/controllers.articles.js [hunk 11]
```diff
@@ -44,21 +45,36 @@ exports.getArticleById = (req, res, next) => {
...
+      res.status(201).send({ postedComment });
```
Comment (popatre) [inline controllers/controllers.articles.js]: The key needs to be `comment`

File: controllers/controllers.errors.js [hunk 12]
```diff
@@ -11,11 +11,13 @@ exports.handleCustomErrors = (err, req, res, next) => {
 };
 
 exports.handle400errors = (err, req, res, next) => {
-  if (err.code === "22P02") {
+  if (err.code) {
```
Comment (popatre) [inline controllers/controllers.errors.js]: I think this was better before tbh - means you can handle different codes differently,

File: models/models.articles.js [hunk 13]
```diff
@@ -68,3 +67,24 @@ exports.checkIfArticleExists = (articleId) => {
     }
   });
 };
+
+exports.insertCommentByArticleId = (articleId, reqBody) => {
+  const { body, username } = reqBody;
+
+  if (!username || !body) {
```
Comment (popatre) [inline models/models.articles.js]: Nice this !

**Review by popatre — CHANGES_REQUESTED** [review]
Could you look at the comments, then resubmit once changed?

File: __tests__/app.test.js (hunk 1 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: __tests__/app.test.js (hunk 2 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: __tests__/app.test.js (hunk 3 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: __tests__/app.test.js (hunk 4 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: controllers/controllers.errors.js (hunk 12 above)
Comment (jaketysome) [reply controllers/controllers.errors.js]: Done

File: __tests__/app.test.js (hunk 5 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: __tests__/app.test.js (hunk 6 above)
Comment (jaketysome) [reply __tests__/app.test.js]: Done

File: controllers/controllers.articles.js (hunk 11 above)
Comment (jaketysome) [reply controllers/controllers.articles.js]: Done

File: controllers/controllers.articles.js (hunk 10 above)
Comment (jaketysome) [reply controllers/controllers.articles.js]: Done

**Pull request merged**